packages/client-sdk-nodejs/src/internal/grpc/client-timeout-interceptor.ts

```diff
@@ -1,14 +1,123 @@
+        );
+        options.deadline = deadline;
+      }
+      return new InterceptingCall(nextCall(options));
+    };
+  }
+}
+
+class RetryUntilTimeoutInterceptor {
+  private readonly logger: MomentoLogger;
+  private readonly responseDataReceivedTimeoutMs: number;
+  private readonly overallRequestTimeoutMs: number;
+  private overallDeadline?: Date;
+
+  constructor(
+    loggerFactory: MomentoLoggerFactory,
+    overallRequestTimeoutMs: number,
+    responseDataReceivedTimeoutMs: number
+  ) {
+    this.logger = loggerFactory.getLogger(this);
+    this.responseDataReceivedTimeoutMs = responseDataReceivedTimeoutMs;
+    this.overallRequestTimeoutMs = overallRequestTimeoutMs;
+  }
+
+  public createTimeoutInterceptor(): Interceptor {
+    return (options, nextCall) => {
+      // Replace default timeout with the desired overall timeout
+      // on the first time through and set the first incremental timeout
+      if (this.overallDeadline === undefined) {
+        const newDate = new Date(Date.now());
+        newDate.setMilliseconds(
+          newDate.getMilliseconds() + this.overallRequestTimeoutMs
+        );
+        this.overallDeadline = newDate;
+
+        const deadline = new Date(Date.now());
+        deadline.setMilliseconds(
+          deadline.getMilliseconds() + this.responseDataReceivedTimeoutMs
+        );
+        options.deadline = deadline;
+        this.logger.debug(`new deadline set to ${options.deadline.valueOf()}`);
+        return new InterceptingCall(nextCall(options));
+      }
+
+      // If the received deadline is equal to the overall deadline, we've
+      // maxed out the retries and should cancel the request.
+      const receivedDeadline = options.deadline;
+      if (
+        receivedDeadline !== undefined &&
+        receivedDeadline === this.overallDeadline
+      ) {
+        this.logger.debug(
+          'Unable to successfully retry request within overall timeout, canceling request'
+        );
+        // reset overall deadline for next request
+        this.overallDeadline = undefined;
+        const call = new InterceptingCall(nextCall(options));
+        call.cancelWithStatus(
+          status.CANCELLED,
+          'Unable to successfully retry request within overall timeout'
+        );
+        return call;
+      }
+
+      // Otherwise, we've hit an incremental timeout and must set the next deadline.
+      const newDeadline = new Date(Date.now());
+      newDeadline.setMilliseconds(
+        newDeadline.getMilliseconds() + this.responseDataReceivedTimeoutMs
+      );
+      if (newDeadline > this.overallDeadline) {
+        this.logger.debug(
+          `new deadline would exceed overall deadline, setting to overall deadline ${this.overallDeadline.valueOf()}`
+        );
+        options.deadline = this.overallDeadline;
+      } else {
+        this.logger.debug(`new deadline set to ${newDeadline.valueOf()}`);
+        options.deadline = newDeadline;
+      }
```

Comment:
This is almost there. We need to make a couple small changes:

- get the current time
- if the `overallDeadline` has already elapsed (current time exceeds `overallDeadline`), then cancel;
- otherwise we have *some* time left to run the operation.
    - In that case we set the deadline to current ms + `responseDataReceivedTimeoutMs`
    - We clip the deadline to `overallDeadline` if needed

I recommend a helper function to compute this since the calculation is repeated on line 76. We should also document we have intentionally decided to run the operation even when we have less time than `responseDataReceivedTimeoutMs`